opentelemetry-api/src/opentelemetry/context/__init__.py

```diff
@@ -12,141 +12,123 @@
+    return get_current().copy()
+
+
+def create_key(key: str) -> "object":
+    # FIXME Implement this
+    raise NotImplementedError
+
+
+def get_value(key: str, context: typing.Optional[Context] = None) -> "object":
+    """
+    To access the local state of an concern, the Context API
+    provides a function which takes a context and a key as input,
+    and returns a value.
+
+    Args:
+        key: The key of the value to retrieve.
+        context: The context from which to retrieve the value, if None, the current context is used.
+    """
+    if context is not None:
+        return context.get_value(key)
+    return get_current().get_value(key)
+
+
+def set_value(
+    key: str, value: "object", context: typing.Optional[Context] = None
+) -> Context:
+    """
+    To record the local state of a cross-cutting concern, the
+    Context API provides a function which takes a context, a
+    key, and a value as input, and returns an updated context
+    which contains the new value.
+
+    Args:
+        key: The key of the entry to set
+        value: The value of the entry to set
+        context: The context to copy, if None, the current context is used
+    """
+    new_context = _copy_context(context)
+    new_context.set_value(key, value)
+    return new_context
+
+
+def remove_value(
+    key: str, context: typing.Optional[Context] = None
+) -> Context:
+    """
+    To remove a value, this method returns a new context with the key cleared.
+    Note that the removed value still remains present in the old context.
+
+    Args:
+        key: The key of the entry to remove
+        context: The context to copy, if None, the current context is used
+    """
+    new_context = _copy_context(context)
+    new_context.remove_value(key)
+    return new_context
+
+
+def get_current() -> Context:
+    """
+    To access the context associated with program execution,
+    the Context API provides a function which takes no arguments
+    and returns a Context.
+    """
+    global _CONTEXT  # pylint: disable=global-statement
+    if _CONTEXT is None:
+        # FIXME use a better implementation of a configuration manager to avoid having
+        # to get configuration values straight from environment variables
+        _available_contexts = {}  # typing.Dict[str, Context]
+
+        for entry_point in iter_entry_points("opentelemetry_context"):
+            try:
+                _available_contexts[entry_point.name] = entry_point.load()  # type: ignore
+            except Exception as err:  # pylint: disable=broad-except
+                logger.warning(
+                    "Could not load entry_point %s:%s", entry_point.name, err
+                )
+
+        configured_context = environ.get(
+            "OPENTELEMETRY_CONTEXT", "default_context"
+        )  # type: str
+        _CONTEXT = _available_contexts[configured_context]()  # type: ignore
+    return _CONTEXT  # type: ignore
+
+
+def set_current(context: Context) -> None:
```

Comment:
If this is supposed to represent the current context, then we cant just use a global variable. We need a context-local variable (or at least a thread local variable). But this is already what the context inside does, so it seems we have mixed up "current Context" and "configured context manager".

Reply:
Maybe my interpretation of what `set_current` is intended to achieve is wrong. My understanding is that `set_current` is used here to update the global context which can then be referenced other places in the application with updated values rather than updating the context within a context manager. This I think can be achieved with the global here, though I might be missing something.

Reply:
@Oberon00 , what do you mean with "configured context manager"?

Reply:
I think we have to distinguish three things:
1. A single context object, which is basically just a String -> Object dictionary. There is only at most one value for each key in a particular context. I think we are currently missing that concept.
2. The current value of a particular context variable/key in some context (what `get_value` returns).
3. The context manager, which is responsible for determining what the "current" context is, creating a new context object based on the parent context if a new logical execution context is entered / get_current is called in a logical execution context where no context object exists yet, etc.. This is what we currently call (IMHO confusingly) "Context".
4. The logical execution context, an abstract concept, not a `class` or object. E.g. a thread of execution, the context of a async task, a greenlet, etc. The context manager defines what an execution context is.

Reply:
Some clarifications:
>1. A single context object, which is basically just a String -> Object dictionary. There is only at most one value for each key in a particular context. I think we are currently missing that concept.

This is currently represented by the `Context` interface which is implemented by either `ContextVarsContext` or `ThreadLocalContext` depending on the version of Python

>3. The context manager, which is responsible for determining what the "current" context is, creating a new context object based on the parent context if a new logical execution context is entered / get_current is called in a logical execution context where no context object exists yet, etc.. This is what we currently call (IMHO confusingly) "Context".

This is implemented via the `get_current` & `set_current` methods

Reply:
I was about to comment the same. Regarding point 1, [here](https://github.com/open-telemetry/opentelemetry-python/pull/395/files#diff-0cab0e9700c8aa1a626de6dce9870b75R20) is the class that represents the concept of a single context object. It is implemented [here](https://github.com/open-telemetry/opentelemetry-python/pull/395/files#diff-3b32e623c0b44273ac8c275f6aae28d3R19), [here](https://github.com/open-telemetry/opentelemetry-python/pull/395/files#diff-ff89937cff2c1d9bf9b2c9b3fb4e9a06R28) and [here](https://github.com/open-telemetry/opentelemetry-python/pull/395/files#diff-25c9264fe792bfdb3af457b682c14dceR21).

Regarding your first EDIT, there is nothing named `set_context`. Did you mean `set_current`?

Reply:
@codeboten @ocelotl: No, ContextVarsContext and ThradLocalContext are not a contexts in the OTEP 66 sense. E.g. ThreadLocalContext is a mapping `(thread, key) -> value` instead of just `key -> value`. That's what I wanted to clarify in my EDIT2. In the OTEP, `Propagator.extract` returns a context. Which return type would you use given the types in this PR, so that you can give the context to another thread or execution context and it still keeps its values?

Reply:
I think @Oberon00 has a good point here, and this will not work as expected.

Using his context queue example:

```python
from queue import Queue
import asyncio

from opentelemetry import context
from opentelemetry.sdk.context.contextvars_context import ContextVarsContext

context.set_current(ContextVarsContext())

q = Queue()

async def main():
    q.put(context.get_current().copy())
    q.put(context.get_current().copy())
    asyncio.gather(work(1), work(2))

async def work(val):
    context.set_current(q.get_nowait())
    context.get_current().set_value('k', val)
    # At this point context.get_current() should be the one we popped from the
    # queue, and current_context.get_value('k') should be `val`

await main()
```

This works as expected, but only incidentally. If the call to `set_current` happens in one task in between the `set_current` and `get_current` calls in another, we get some weird behavior. Running effectively the same code, but manipulating the call order with an `Event` shows this:

```python
from queue import Queue
from threading import Event
import asyncio

from opentelemetry import context
from opentelemetry.sdk.context.contextvars_context import ContextVarsContext

context.set_current(ContextVarsContext())

# Helper to print Contexts
def tos(context):
    return "<Context at {}: {}>".format(
        hex(id(context)),
        ','.join(["{}={}".format(k, v.get(None))
                  for k, v in context._contextvars.items()])
    )

q = Queue()
run1 = Event()
run2 = Event()

async def main():
    q.put(context.get_current().copy())
    q.put(context.get_current().copy())
    asyncio.gather(a1(), a2())

async def a1():
    # (Step 1)
    context.set_current(q.get_nowait())
    context.get_current().set_value('k', 1)

    # Expected context is current, k=1
    print("[task 1] {}".format(tos(context.get_current())))

    # (GOTO 2)
    run2.set(); await asyncio.sleep(0); run1.wait();

    # (Step 3)
    # Current context has changed AND k=None because the call to `set` k=2
    # happened in the other task!
    print("[task 1] {}".format(tos(context.get_current())))


async def a2():
    run2.wait()

    # (Step 2)
    context.set_current(q.get_nowait())
    context.get_current().set_value('k', 2)

    # Expected context is current, k=2
    print("[task 2] {}".format(tos(context.get_current())))

    # (GOTO 3)
    run1.set(); await asyncio.sleep(0);

await main()
```
This prints:
```
[task 1] <Context at 0x110349dd0: k=1>
[task 2] <Context at 0x110349250: k=2>
[task 1] <Context at 0x110349250: k=None>
```

Task 1 gets the context from the `set_current` call in task 2, but doesn't get its context variable values because `set` calls are scoped to the task! You could show roughly the same thing with threads and threadlocals.

(Sorry for the long example here)

Reply:
@ocelotl Yes, sorry I meant `set_current`.